File: 0133-clone-graph/0133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if(node==null){
            return null;
        }
        //Original Node, Clone Node
        Map<Node, Node> map = new HashMap<>();
        Queue<Node> queue = new ArrayDeque<>();
        queue.add(node);
        map.put(node, new Node(node.val, new ArrayList<>()));
        while(!queue.isEmpty()){
            Node curr = queue.poll();
            for(Node neighbor: curr.neighbors){
                if(!map.containsKey(neighbor)){
                    map.put(neighbor, new Node(neighbor.val, new ArrayList<>()));
                    queue.add(neighbor);
                }
                Node cloneNode = map.get(curr);
                Node neighborNode = map.get(neighbor);
                cloneNode.neighbors.add(neighborNode);
            }
        }
        return map.get(node);
    }
}

File: 0133-clone-graph/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if(node==NULL){
            return NULL;
        }
        //Original Node, Clone Node
        unordered_map<Node*, Node*> map;
        queue<Node*> q;
        q.push(node);
        map[node] = new Node(node->val);
        while(!q.empty()){
            Node* curr = q.front();
            q.pop();
            for(Node* neighbor: curr->neighbors){
                if(map.find(neighbor)==map.end()){
                    map[neighbor] = new Node(neighbor->val);
                    q.push(neighbor);
                }
                Node* cloneNode = map[curr];
                Node* neighborNode = map[neighbor];
                cloneNode->neighbors.push_back(neighborNode);
            }
        }
        return map[node];
    }
};